### PR TITLE
Implement the feature that close issuse as archived/merged/resolved

### LINKED ISSUE #22793: [Feature] Add reason for closed issues (duplicate, archived etc)
### Feature Description

Instead of using labels to describe issue states (dupe, archived, etc..), additional descriptive metadata could be added to issues.

To ensure backwards compatibility, open/closed could be kept, but then you could add search by state, etc.. We have "mentioned in issues #xyz" issue linking, and we could leverage that to be "duplicate issue #abc marked as duplicate of this issue. The subscribers to the initial ticket that is marked as dupe could be transferred to the open ticket.

```go
type IssueState int64
const (
  Archived IssueState = iota
  Duplicate
  Draft   // Issue that is open but still being worked on, similar to draft PR. We could also use this for draft PRs.
  // ...
) 

```

### Screenshots

screenshot of github allowing to close issue with a specific state

### PR DESCRIPTION
**Pull request by sillyguodong**
Close #22793
Implement the feature of "Close as archived / resolved / merged"

1. Create a new issues, open the dropdown, user can select what they want to do.
image

2. click the left button, submit with a comment.
image
image

3. in the issue list, we can see what kind of `closed status` in the tooltip
image

### CODE REVIEW

File: models/issues/issue.go [hunk 1]
```diff
@@ -2498,3 +2498,20 @@ func DeleteOrphanedIssues(ctx context.Context) error {
...
 	return issue.OriginalAuthor != "" && issue.OriginalAuthorID != 0
 }
+
+type IssueCloseState int
+
+const (
+	// IssueStateOpen
+	IssueStateOpen IssueCloseState = iota
+	// IssueStateLocked
+	IssueStateLocked
```
Comment (wxiaoguang) [inline models/issues/issue.go]: I am not sure whether this is the final design (I see this PR is WIP), but I'd like to have quick comment about this design: I guess the "Locked" should be a separate state. For example: Merged, Merged+Locked

File: templates/repo/issue/view_content.tmpl [hunk 2]
```diff
@@ -108,18 +108,53 @@
...
 												{{.locale.Tr "repo.issues.reopen_issue"}}
 											</button>
 										{{else}}
-											<button id="status-button" class="ui red basic button" tabindex="6" data-status="{{.locale.Tr "repo.issues.close_issue"}}" data-status-and-comment="{{.locale.Tr "repo.issues.close_comment_issue"}}" data-status-val="close">
+											<button id="status-button" type="submit" class="ui basic button" data-status="{{.locale.Tr "repo.issues.close_issue"}}" data-status-and-comment="{{.locale.Tr "repo.issues.close_comment_issue"}}" data-status-val="close">
 												{{.locale.Tr "repo.issues.close_issue"}}
 											</button>
 										{{end}}
+
+										<div id="state-dropdown" class="ui dropdown jump item tooltip gt-px-3" tabindex="-1" data-content="">
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content.tmpl]: I guess you need this https://fomantic-ui.com/modules/dropdown.html#floating , instead of re-inventing wheels

Comment (sillyguodong) [reply templates/repo/issue/view_content.tmpl]: yes, that's exactly what i need😂

File: web_src/js/features/aria.js [hunk 3]
```diff
@@ -97,8 +99,9 @@ function attachOneDropdownAria($dropdown) {
     // if the popup is visible and has an active/selected item, use its id as aria-activedescendant
     if (menuVisible) {
       $focusable.attr('aria-activedescendant', $active.attr('id'));
-    } else if (!isComboBox) {
-      // for menu, when the popup is hidden, no need to keep the aria-activedescendant, and clear the active/selected item
+    } else if (!isComboBox && !needKeepAtive) {
+      // for menu in most cases, when the popup is hidden, no need to keep the aria-activedescendant, and clear the active/selected item
+      // if you really need to keep the aria-activedescendant, plz add a className 'keep-active' to the dropdown
```
Comment (wxiaoguang) [inline web_src/js/features/aria.js]: `menu` doesn't keep last selected item, and it would break some screen readers IIRC.

You need to make it a `combobox`.


For example: `class=js-aria-comobox`, then `isCombobox=true`

File: web_src/js/features/aria.js [hunk 4]
```diff
@@ -38,6 +38,8 @@ function attachOneDropdownAria($dropdown) {
   const listPopupRole = isComboBox ? 'listbox' : 'menu';
   const listItemRole = isComboBox ? 'option' : 'menuitem';
 
+  const needKeepAtive = $dropdown.hasClass('keep-active');
```
Comment (sillyguodong) [inline web_src/js/features/aria.js]: @wxiaoguang hi, in this case, i need to keep `check` icon active even if the menu is hidden and reopened.
So, I add a classname 'keep-active' to the dropdown. But I'm not sure that's a good idea. Do you have better one?
image

Comment (wxiaoguang) [reply web_src/js/features/aria.js]: I have commented above 😂

File: web_src/js/features/aria.js (hunk 3 above)
Comment (sillyguodong) [reply web_src/js/features/aria.js]: wow, a quick comment~That's exactly what I wanted to ask you.

File: web_src/js/features/repo-issue.js [hunk 5]
```diff
@@ -226,18 +226,43 @@ export function initRepoIssueCodeCommentCancel() {
...
-    $('#status').val($statusButton.data('status-val'));
     $('#comment-form').trigger('submit');
   });
 }
 
+export function initRepoIssueStatusDropdown() {
+  const $statusDropdown = $('#status-dropdown');
+  const $statusMenu = $statusDropdown.find('> .menu');
+  $statusMenu.find('> .item').each((_, item) => {
+    $(item).on('click', (e) => {
```
Comment (wxiaoguang) [inline web_src/js/features/repo-issue.js:249]: I guess a lot of code could be simplified by Dropdown's builtin mechisam

The dropdown could be used as a `<select>`-like component, it can have an internal `input`, it can remember it's internal value and update the UI by the internal value.

File: web_src/js/modules/aria/dropdown.js [hunk 6]
```diff
@@ -140,7 +140,7 @@ function attachInit($dropdown) {
   // There are 2 possible solutions about the role: combobox or menu.
   // The idea is that if there is an input, then it's a combobox, otherwise it's a menu.
   // Since #19861 we have prepared the "combobox" solution, but didn't get enough time to put it into practice and test before.
-  const isComboBox = $dropdown.find('input').length > 0;
+  const isComboBox = $dropdown.find('input').length > 0 || $dropdown.hasClass('js-aria-comobox');
```
Comment (wxiaoguang) [inline web_src/js/modules/aria/dropdown.js]: I guess you do not need this.

File: web_src/css/repository.css [hunk 7]
```diff
@@ -333,7 +333,7 @@
   margin-right: 3px;
 }
 
-.repository.file.list #repo-files-table td > * {
+.repository.file.list #repo-files-table td>* {
```
Comment (wxiaoguang) [inline web_src/css/repository.css]: Too many unrelated changes in this file.

Comment (sillyguodong) [reply web_src/css/repository.css]: Sorry, I have restored some unexpected css format now

File: web_src/js/modules/aria/dropdown.js (hunk 6 above)
Comment (sillyguodong) [reply web_src/js/modules/aria/dropdown.js]: You are right, I misunderstood the suggestion you gave me

File: models/issues/issue.go [hunk 8]
```diff
@@ -2498,3 +2519,20 @@ func DeleteOrphanedIssues(ctx context.Context) error {
...
 	return issue.OriginalAuthor != "" && issue.OriginalAuthorID != 0
 }
+
+type IssueClosedStatus int64
+
+const IssueClosedStatusUndefined = -1
+
+const (
+	// IssueClosedStatusOpen
+	IssueClosedStatusOpen IssueClosedStatus = iota
```
Comment (a1012112796) [inline models/issues/issue.go:2529]: looks meanless, can check it by `Issue.IsClosed`

File: models/issues/issue.go [hunk 9]
```diff
@@ -2498,3 +2519,20 @@ func DeleteOrphanedIssues(ctx context.Context) error {
...
+	// IssueClosedStatusCommonClosed
+	IssueClosedStatusCommonClosed
+	// IssueClosedStatusArchived
+	IssueClosedStatusArchived
+	// IssueClosedStatusResolved
+	IssueClosedStatusResolved
+	// IssueClosedStatusMerged
+	IssueClosedStatusMerged
```
Comment (a1012112796) [inline models/issues/issue.go:2537]: how about add `duplicate` and `stale` status?

File: models/issues/comment.go [hunk 10]
```diff
@@ -130,6 +130,12 @@ const (
 	CommentTypePRScheduledToAutoMerge
 	// 35 pr was un scheduled to auto merge when checks succeed
 	CommentTypePRUnScheduledToAutoMerge
+	// 36
+	CommentTypeCloseAsArchived
+	// 37
+	CommentTypeCloseAsResolved
+	// 38
+	CommentTypeCloseAsMerged
```
Comment (wxiaoguang) [inline models/issues/comment.go:138]: I guess it's better to extend the `CommentTypeClose` , instead of adding too many new comment types.

**Comment by sillyguodong** [comment]
This PR is not good enough, I am trying to rewrite this in a new branch, and will create a new PR replace this when it was done. So close it for now.
